website/static/js/nodesPrivacy.js

```diff
@@ -0,0 +1,272 @@
+     */
+    var i;
+    var nodeId = nodeTree.node.id;
+    nodesOriginal[nodeId] = {
+        public: nodeTree.node.is_public,
+        id: nodeTree.node.id,
+        addons: nodeTree.node.addons,
+        title: nodeTree.node.title,
+        canWrite: nodeTree.node.can_write,
+        changed: false
+    };
+
+    if (nodeTree.children) {
+        for (i in nodeTree.children) {
+            nodesOriginal = getNodesOriginal(nodeTree.children[i], nodesOriginal);
+        }
+    }
+    return nodesOriginal;
+}
+
+function patchNodesPrivacy(nodes) {
+    /**
+     * patches all the nodes in a changed state
+     * uses API v2 bulk requests
+     */
+    var nodesPatch = [];
+    var promise;
+    nodesPatch = nodes.splice(0,10).map(function (node) {
+        return {
+            'type': 'nodes',
+            'id': node.id,
+            'attributes': {
+                'public': node.public
+            }
+        };
+    });
+    return $.ajax({
+        url: API_BASE,
+        type: 'PATCH',
+        dataType: 'json',
+        contentType: 'application/vnd.api+json; ext=bulk',
+        crossOrigin: true,
+        xhrFields: {withCredentials: true},
+        processData: false,
+        data: JSON.stringify({
+            data: nodesPatch
+        })
+    }).done(function (response) {
+        if (nodes.length === 0) {
+            return;
+        }
+        return patchNodesPrivacy(nodes);
+    });
+}
+
+var NodesPrivacyViewModel = function(data, parentIsPublic) {
+    /**
+     * view model which corresponds to nodes_privacy.mako (#nodesPrivacy)
+     *
+     * @type {NodesPrivacyViewModel}
+     */
+    var self = this;
+    var nodesOriginal = {};
+    var treebeardUrl = data.node.api_url  + 'get_node_tree/';
+    self.loadingResults = ko.observable(false);
+    //state of current nodes
+    self.nodesState = ko.observableArray();
+    self.nodesState.subscribe(function(newValue) {
+        m.redraw(true);
+    });
+    //original node state on page load
+    self.nodesOriginal = ko.observableArray();
+    self.nodesOriginal.subscribe(function(newValue) {
+    });
+
+    $('#nodesPrivacy').on('hidden.bs.modal', function () {
+        self.clear();
+    });
+
+    self.loadingResults(true); // enables spinner
+    /**
+     * get node tree for treebeard from API V1
+     */
+    $.ajax({
+        url: treebeardUrl,
+        type: 'GET',
+        dataType: 'json'
+    }).done(function(response) {
+        nodesOriginal = getNodesOriginal(response[0], nodesOriginal);
+        self.nodesOriginal(nodesOriginal);
+        var nodesState = $.extend(true, {}, nodesOriginal);
+        var nodeParent = response[0].node.id;
+        //change node state to reflect button push by user on project page (make public | make private)
+        nodesState[nodeParent].public = !parentIsPublic;
+        nodesState[nodeParent].changed = true;
+        self.nodesState(nodesState);
+        new NodesPrivacyTreebeard(response, self.nodesState, nodesOriginal);
```

Comment:
Separating your data fetching into a method will allow you to separate treebeard from this class altogether. In your `NodesPrivacy` class, you would be able to do

```
self.viewModel = new NodesPrivacyViewModel(self.data, parentNodePrivacy);
self.viewModel.fetch()
    .done(function(response) {
        new NodesPrivacyTreebeard(...);
    })
```

This is advantageous because `NodesPrivacyViewModel` will no longer depend on the DOM (the treebeard expects an HTML tag with the `#grid` ID), which makes testing much easier.